### PR TITLE
[AD-254] Implement FLOOR for DATETIME types.

### PR DESCRIPTION
**Pull request by birschick-bq**
### Summary

[AD-254] Implement FLOOR for DATETIME types.

### Description

Implements `FLOOR (<datetime> TO <timeunit>)`
- Uses equivalent of `dateadd(<timeunit>,datediff(<timeunit>,0,@datetime),0)` (https://stackoverflow.com/questions/85373/floor-a-date-in-sql-server/85379#85379)
- Except for YEAR and MONTH which use `$dateToString` and `$dateFromString`
- Implements an emulated version of integer division using `$mod`.

### Related Issue

https://bitquill.atlassian.net/browse/AD-254

### Additional Reviewers
@birschick-bq
@mitchell-elholm
@alexey-temnikov 
@andiem-bq

### CODE REVIEW

File: src/main/java/software/amazon/documentdb/jdbc/DocumentDbQueryExecutor.java [hunk 1]
```diff
@@ -135,19 +134,19 @@ protected void cancelQuery() throws SQLException {
...
                 } else {
-                    resetQueryState();
                     throw SqlError.createSQLException(
                             LOGGER,
                             SqlState.CONNECTION_EXCEPTION,
                             SqlError.QUERY_FAILED, e);
                 }
             }
+        } finally {
+            resetQueryState();
```
Comment (birschick-bq) [inline src/main/java/software/amazon/documentdb/jdbc/DocumentDbQueryExecutor.java:149]: After an exception, the state is not always consistent.

File: calcite-adapter/src/test/java/software/amazon/documentdb/jdbc/query/DocumentDbQueryMappingServiceTest.java [hunk 2]
```diff
@@ -1434,33 +1434,17 @@ void testDateFunctions() throws SQLException {
...
+            BsonDocument.parse(
+                    "{\"$addFields\":"
+                    + " {\"EXPR$0\": {\"$add\": [\"$field\", {\"$multiply\": [604800000, 1]}]},"
+                    + " \"EXPR$1\": {\"$add\": [\"$field\", {\"$multiply\": [86400000, 2]}]},"
+                    + " \"EXPR$2\": {\"$add\": [\"$field\", {\"$multiply\": [3600000, 3]}]},"
+                    + " \"EXPR$3\": {\"$add\": [\"$field\", {\"$multiply\": [60000, 4]}]},"
+                    + " \"EXPR$4\": {\"$add\": [\"$field\", {\"$multiply\": [1000, 5]}]},"
+                    + " \"EXPR$5\": {\"$add\": [\"$field\", {\"$divide\":"
+                        + " [{\"$subtract\": [{\"$multiply\": [1, 6]},"
+                            + " {\"$mod\": [{\"$multiply\": [1, 6]}, 1000]}]}, 1000]}]}}}").toJson(),
```
Comment (birschick-bq) [inline calcite-adapter/src/test/java/software/amazon/documentdb/jdbc/query/DocumentDbQueryMappingServiceTest.java]: Integer divide.

File: calcite-adapter/src/test/java/software/amazon/documentdb/jdbc/query/DocumentDbQueryMappingServiceTest.java [hunk 3]
```diff
@@ -1520,26 +1504,12 @@ void testDateFunctions() throws SQLException {
...
+                        + " \"EXPR$1\": {\"$divide\": [{\"$subtract\": [{\"$subtract\": [\"$field\", \"$field\"]}, {\"$mod\": [{\"$subtract\": [\"$field\", \"$field\"]}, 86400000]}]}, 86400000]},"
+                        + " \"EXPR$2\": {\"$divide\": [{\"$subtract\": [{\"$subtract\": [\"$field\", \"$field\"]}, {\"$mod\": [{\"$subtract\": [\"$field\", \"$field\"]}, 3600000]}]}, 3600000]},"
+                        + " \"EXPR$3\": {\"$divide\": [{\"$subtract\": [{\"$subtract\": [\"$field\", \"$field\"]}, {\"$mod\": [{\"$subtract\": [\"$field\", \"$field\"]}, 60000]}]}, 60000]},"
+                        + " \"EXPR$4\": {\"$divide\": [{\"$subtract\": [{\"$subtract\": [\"$field\", \"$field\"]}, {\"$mod\": [{\"$subtract\": [\"$field\", \"$field\"]}, 1000]}]}, 1000]},"
+                        + " \"EXPR$5\": {\"$multiply\": [{\"$divide\": [{\"$subtract\": [{\"$subtract\": [\"$field\", \"$field\"]}, {\"$mod\": [{\"$subtract\": [\"$field\", \"$field\"]}, 1000]}]}, 1000]}, 1000000]}}}"),
```
Comment (birschick-bq) [inline calcite-adapter/src/test/java/software/amazon/documentdb/jdbc/query/DocumentDbQueryMappingServiceTest.java]: Now includes integer divide.

File: src/test/java/software/amazon/documentdb/jdbc/DocumentDbStatementBasicTest.java [hunk 4]
```diff
@@ -428,12 +429,12 @@ void testQuerySumOne() throws SQLException {
...
     void testQueryTimestampAdd() throws SQLException {
         final String tableName = "testTimestampAdd";
-        final long dateTime = Instant.parse("2020-01-01T00:00:00.00Z").toEpochMilli();
-        final long weekAfterDateTime = Instant.parse("2020-01-08T00:00:00.00Z").toEpochMilli();
-        final long dayAfterDateTime = Instant.parse("2020-01-02T00:00:00.00Z").toEpochMilli();
-        final long hourAfterDateTime =  Instant.parse("2020-01-01T01:00:00.00Z").toEpochMilli();
-        final long minuteAfterDateTime = Instant.parse("2020-01-01T00:01:00.00Z").toEpochMilli();
-        final long secondAfterDateTime =  Instant.parse("2020-01-01T00:00:01.00Z").toEpochMilli();
+        final long dateTime = Instant.parse("2020-02-22T00:00:00.00Z").toEpochMilli();
+        final long weekAfterDateTime = Instant.parse("2020-02-29T00:00:00.00Z").toEpochMilli();
```
Comment (birschick-bq) [inline src/test/java/software/amazon/documentdb/jdbc/DocumentDbStatementBasicTest.java:433]: Wanted to test that leap year was being handled.

File: src/test/java/software/amazon/documentdb/jdbc/DocumentDbStatementBasicTest.java [hunk 5]
```diff
@@ -447,13 +448,22 @@ void testQueryTimestampAdd() throws SQLException {
 
         // Add 1 day to a date column.
         final ResultSet resultSet = statement.executeQuery(
-                String.format("SELECT TIMESTAMPADD(DAY"
+                String.format("SELECT TIMESTAMPADD(WEEK"
```
Comment (birschick-bq) [inline src/test/java/software/amazon/documentdb/jdbc/DocumentDbStatementBasicTest.java:451]: Added test for `WEEK`.

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbRules.java [hunk 6]
```diff
@@ -579,6 +590,101 @@ public static String translateMonthName(final RexCall rexCall, final List<String
...
+                            + " {'date': %1$s, 'format': '%2$s-%3$s-01T00:00:00Z'}}}}",
+                    strings.get(0), yearFormat, monthFormat);
+        }
+
+        private static String formatNumericFloorOperation(
+                final List<String> strings,
+                final TimeUnitRange timeUnitRange) throws SQLFeatureNotSupportedException {
+
+            final Instant baseDate = timeUnitRange == TimeUnitRange.WEEK
+                    ? Instant.parse("1970-01-05T00:00:00Z") // Monday (or first day of week)
```
Comment (birschick-bq) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbRules.java]: Needs to be a date on a Monday (or which ever first day of the week you need).

Comment (alexey-temnikov) [reply calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbRules.java]: Should we use constant to replace `Instant.parse("1970-01-05T00:00:00Z")`, so we don't need to parse string every time?

Comment (birschick-bq) [reply calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbRules.java]: Fixed.

File: docker/TableauConnector/dialect.tdd [hunk 7]
```diff
@@ -583,6 +588,14 @@
     <aggregation value='AGG_SECOND'/>
     <aggregation value='AGG_ISO_YEAR'/>
     <aggregation value='AGG_ISO_WEEKDAY'/>
+    <aggregation value='TRUNC_YEAR'/>
```
Comment (birschick-bq) [inline docker/TableauConnector/dialect.tdd:591]: Added support for aggregating FLOOR/DATETRUNC function.

**Review by andiemontoyeah — APPROVED** [review]
lgtm!

**Review by mitchell-elholm — APPROVED** [review]

**Pull request merged**